interfaces/dotnet/Cantera/src/SpeciesCollection.cs

```diff
@@ -0,0 +1,206 @@
+        _species.GetEnumerator();
+
+    IEnumerator IEnumerable.GetEnumerator() =>
+        _species.GetEnumerator();
+
+    public int IndexOf(Species species) =>
+        IndexOf(species.Name);
+
+    public int IndexOf(string name)
+    {
+        if (_speciesIndexByName.Value.TryGetValue(name, out var index))
+        {
+            return index;
+        }
+
+        return -1;
+    }
+
+    public bool Contains(Species species) =>
+        Contains(species.Name);
+
+    public bool Contains(string name) =>
+        _speciesIndexByName.Value.ContainsKey(name);
+
+    public void SetMassFractions(params (string speciesName, double fraction)[] tuples)
+    {
+        SetMassFractions((IEnumerable<(string, double)>)tuples);
+    }
+
+    public void SetMassFractions(
+        IEnumerable<(string speciesName, double fraction)> tuples)
+    {
+        var fractions = new double[_species.Count];
+
+        foreach(var tuple in tuples)
+            fractions[EnsureIndexOf(tuple.speciesName)] = tuple.fraction;
+
+        MassFractions = fractions;
+    }
+
+    public void SetMassFractions(params (Species species, double fraction)[] tuples) =>
+        SetMassFractions((IEnumerable<(Species, double)>) tuples);
+
+    public void SetMassFractions(IEnumerable<(Species species, double fraction)> tuples)
+    {
+        var fractions = new double[_species.Count];
+
+        foreach(var tuple in tuples)
+            fractions[EnsureIndexOf(tuple.species)] = tuple.fraction;
+
+        MassFractions = fractions;
+    }
+
+    public void SetMoleFractions(params (string speciesName, double fraction)[] tuples)
+    {
+        SetMoleFractions((IEnumerable<(string, double)>)tuples);
+    }
+
+    public void SetMoleFractions(
+                              IEnumerable<(string speciesName, double fraction)> tuples)
+    {
+        var fractions = new double[_species.Count];
+
+        foreach(var tuple in tuples)
+            fractions[EnsureIndexOf(tuple.speciesName)] = tuple.fraction;
+
+        MoleFractions = fractions;
+    }
+
+    public void SetMoleFractions(params (Species species, double fraction)[] tuples) =>
+        SetMoleFractions((IEnumerable<(Species, double)>) tuples);
+
+    public void SetMoleFractions(IEnumerable<(Species species, double fraction)> tuples)
+    {
+        var fractions = new double[_species.Count];
+
+        foreach(var tuple in tuples)
+            fractions[EnsureIndexOf(tuple.species)] = tuple.fraction;
+
+        MoleFractions = fractions;
+    }
+
+    int EnsureIndexOf(Species species) =>
+        EnsureIndexOf(species.Name);
+
+    int EnsureIndexOf(string name)
+    {
+        if (_speciesIndexByName.Value.TryGetValue(name, out var index))
+        {
+            return index;
+        }
+
+        throw new InvalidOperationException(
+            $"Unknown species! {name} is not present in this collection.");
+    }
+
+    public void SetUnormalizedMassFractions(double[] fractions)
```

Comment:
```suggestion
    public void SetUnnormalizedMassFractions(double[] fractions)
```
Likewise for mole fractions below.